tests/Service/FilterFactoryTest.php

```diff
@@ -0,0 +1,89 @@
+<?php declare(strict_types=1);
+
+namespace Lmc\ApiFilter\Service;
+
+use Lmc\ApiFilter\AbstractTestCase;
+use Lmc\ApiFilter\Entity\Value;
+use Lmc\ApiFilter\Filter\FilterIn;
+use Lmc\ApiFilter\Filter\FilterInterface;
+use Lmc\ApiFilter\Filter\FilterWithOperator;
+
+class FilterFactoryTest extends AbstractTestCase
+{
+    private const COLUMN = 'column';
+
+    /** @var FilterFactory */
+    private $filterFactory;
+
+    protected function setUp(): void
+    {
+        $this->filterFactory = new FilterFactory();
+    }
+
+    /**
+     * @param mixed $rawValue of type <T>
+     * @param mixed $expectedValue of type <T>
+     *
+     * @test
+     * @dataProvider provideFilters
+     */
+    public function shouldCreateFilter(
+        string $filter,
+        string $expectedTitle,
+        $rawValue,
+        $expectedValue,
+        string $expectedFilterClass
+    ): void {
+        $result = $this->filterFactory->createFilter(self::COLUMN, $filter, new Value($rawValue));
+
+        $this->assertInstanceOf(FilterInterface::class, $result);
+        $this->assertSame(self::COLUMN, $result->getColumn());
+        $this->assertSame($expectedValue, $result->getValue()->getValue());
+        $this->assertSame($expectedTitle, $result->getTitle());
+        $this->assertInstanceOf($expectedFilterClass, $result);
+    }
+
+    public function provideFilters(): array
+    {
+        return [
+            // filter, expectedTitle, rawValue, expectedValue, expectedFilterClass
+            'eq' => ['eq', self::COLUMN . '_eq', 'value', 'value', FilterWithOperator::class],
```

Comment:
I don't think its necessary to pass the rawValue, if it is the same in all data sets.

Reply:
`title` is computed value so it should definitely be provided by provider. On the otherhand `rawValue` is same all the time and it adds "complexity" to provider.

I don`t think it hides value. If you move it to testcase you know where did it come from....from testcase. And you are saying that it would be same in all cases.

Reply:
Rule of thumb: use data provider to provide only input values which are different for each testcase. Otherwise you ends with quite a complex and hard to read data providers.

You could put there also expected values `FilterInterface::class` and `self::COLUMN` which are hardcoed in the assertions. Which would be obviously pointless.

There must be some balance between DRY and readability. Because of course, the most transparent would be to have the same testcase - just with different expected value - copied for each of these datasets. But this would be hard to maintain and contain lot of repetitive code and even same values. And providing the same value for all datasets is IMHO not worth it, it does not significantly improve the readability.

Reply:
> On the otherhand rawValue is same all the time and it adds "complexity" to provider.

Not all values are the same and thats why they are in the provider - look at `IN` filter value - which is different - because it is normalized from `scalar` to `array`.

But I've tried to simplified that provider..